### DISCUSSION TITLE
Programming in Rust

### DISCUSSION THREAD

**Discussion by Wenn0101** [sparkfun/Arduino_Apollo3#419] [Ideas]
from jerabaul29 in [#408:](https://github.com/sparkfun/Arduino_Apollo3/issues/408)

> This is quite meta / discussion.
> 
> I have had colleagues trying to push me to have a look at Rust. For quite some time I was reluctant ("why would I learn yet another language to do the same stuff I already do in C++?"). However, after some months of advertising by them, some time spent looking at weird memory bugs, etc, I am giving it a look now on my spare time. Actually, I start to get more and more sold to it too ^^ .
> 
> My question is: are there possibilities to / have any user experience with / any suggestions regarding / programming the Artemis platform by writing some Rust instead of some C++? Any thoughts?

**Comment by Wenn0101** [comment]
I have also had quite a few people suggest to me that I should start learning and using Rust.

I have not made any steps to add support to the Artemis. I think a big hurdle would be that the HAL that we use that is provided by ambiq is written in C++ and it may be painful to recreate for Rust.

I wonder how many people would be interested in this.

**Comment by jerabaul29** [comment]
Looks like the rust embedded book has some example with precisely a Cortex-M4F: https://docs.rust-embedded.org/book/start/hardware.html

I wonder if there may be a way at first to just have "a bit of user rust" running on the board that is controlled by the C++ core we have now.

**Comment by gauteh** [comment]
Hi. I have started programming the Artemis Redboard in Rust, and it is now blinking!

I have basically started two approaches:

1) I linked in the ambiq_artemis_sdk as modifed by sparkfun and use the C-api through FFI (https://github.com/gauteh/ambiq-rs/tree/main/ambiq-hal-sys). And that works after some linker magic! It is blinking.


2) I started a native rust hal for the ambiq: https://github.com/gauteh/ambiq-rs/tree/main/ambiq-hal. This imlements embedded-hal through the appropriate pac crate (https://github.com/gauteh/ambiq-apollo3-pac) as well as the generic cortex-m crate.

So far I've only implemented Delay, but that works just fine as well. Mixing both works, but you loose out on the guarantees you otherwise get through the PAC + HAL crate. It is also possible to hack up the embedded-hal traits so using the C hal to just get going quickly, but this won't enforce any of the guarantees either. In other words, you can write rust but regarding the ambiq hal you don't have any more safety than in C. Depending on how much time I want to spend on this I will extend the rust-native traits or just rely on the C hal. 

Anyway; it is running rust now!

The actual crate that can be flashed onto the redboard is: 

  * https://github.com/gauteh/ambiq-rs/tree/main/boards/sparkfun-redboard-artemis-halc
 
or the one that is using partially a rust native HAL: 

  * https://github.com/gauteh/ambiq-rs/tree/main/boards/sparkfun-redboard-artemis

see https://github.com/gauteh/ambiq-rs/tree/main/boards/sparkfun-redboard-artemis-halc for how to build and flash the blink example.

> By the way; Tock OS has some support for the Artemis Redboard Nano. But it doesn't follow the standard embedded-hal ecosystem. But there is some inspiration for register manipulation in rust there.

Regards, Gaute

**Comment by Wenn0101** [comment]
This is very cool! I'll have to see if I can get things blinking with your first approach as well.

Perhaps this will be my motivation to finally get into Rust.

**Reply by gauteh** [reply]
Yeah! docs are ... sparse ... you need cargo-make to run makers (Makefile.toml), as you see there are a bunch of submodules you need to initialize for things to compile.

**Reply by gauteh** [reply]
I split this stuff into a separate repository and updated the example + a simple readme to get blinking. Should be more straightforward (remember to init submodules).

**Comment by gauteh** [comment]
I've now also implemented UART (for the USB-serial port) and I2C.
